### ISSUE TITLE
Data container not compatible with GP models

### ISSUE THREAD

**Issue by fonnesbeck** [pymc-devs/pymc#3500]
Using `Data` containers for input values in GP models fails because functions like `len` and `np.int` are called on the data by various GP operations. For example:

```
TypeError                                 Traceback (most recent call last)
~/repositories/pymc3/pymc3/gp/util.py in infer_shape(X, n_points)
     13         try:
---> 14             n_points = np.int(X.shape[0])
     15         except TypeError:

TypeError: int() argument must be a string, a bytes-like object or a number, not 'TensorVariable'

During handling of the above exception, another exception occurred:
```

It appears to happen  on shape checking, so we should be able to fix this.

**Comment by twiecki** [comment]
For now we should probably check the input and raise an inormative error.

**Comment by bwengals** [comment]
How should this work ideally?  I think replacing `X` if it was in `pm.Data` would currently break things.  Is there a better way to handle shapes?  It would be nice if X could be a `TensorVariable`, and didn't have to be a numpy array.

**Comment by michaelosthege** [comment]
This is fixed in v4 - among other things by #5804.
